sdk/core/Azure.Core/src/Shared/HttpMessageSanitizer.cs

```diff
@@ -1,140 +1,198 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+#nullable enable
 
 using System;
 using System.Collections.Generic;
 using System.Text;
 using System.Linq;
 
-namespace Azure.Core
+namespace Azure.Core;
+
+internal class HttpMessageSanitizer
 {
-    internal class HttpMessageSanitizer
-    {
-        private const string LogAllValue = "*";
-        private readonly bool _logAllHeaders;
-        private readonly bool _logFullQueries;
-        private readonly string[] _allowedQueryParameters;
-        private readonly string _redactedPlaceholder;
-        private readonly HashSet<string> _allowedHeaders;
+    private const string LogAllValue = "*";
+    private readonly bool _logAllHeaders;
+    private readonly bool _logFullQueries;
+    private readonly string[] _allowedQueryParameters;
+    private readonly string _redactedPlaceholder;
+    private readonly HashSet<string> _allowedHeaders;
 
-        internal static HttpMessageSanitizer Default = new HttpMessageSanitizer(Array.Empty<string>(), Array.Empty<string>());
+    [ThreadStatic]
+    private static StringBuilder? s_cachedStringBuilder;
+    private const int MaxCachedStringBuilderCapacity = 1024;
 
-        public HttpMessageSanitizer(string[] allowedQueryParameters, string[] allowedHeaders, string redactedPlaceholder = "REDACTED")
-        {
-            _logAllHeaders = allowedHeaders.Contains(LogAllValue);
-            _logFullQueries = allowedQueryParameters.Contains(LogAllValue);
+    internal static HttpMessageSanitizer Default = new HttpMessageSanitizer(Array.Empty<string>(), Array.Empty<string>());
 
-            _allowedQueryParameters = allowedQueryParameters;
-            _redactedPlaceholder = redactedPlaceholder;
-            _allowedHeaders = new HashSet<string>(allowedHeaders, StringComparer.InvariantCultureIgnoreCase);
-        }
+    public HttpMessageSanitizer(string[] allowedQueryParameters, string[] allowedHeaders, string redactedPlaceholder = "REDACTED")
```

Comment:
While this would be a more controversial change, I wonder if we should consider removing the optional `redactedPlaceholder` and hard code it to `*`.  Doing so would allow us to eliminate the StringBuilder and sanitize using `Span.Slice` exclusively.

Looking through the code base, nobody uses this arg, except in a single test.

Reply:
Who are you asking?
Certainly, if the redacted text is known to consist of a single char, then we can further optimize (although, there's an edge case when a disallowed query param has a zero-length value -- nothing after the equals sign -- and the redacted string could still be longer than the input).
However, without changing the return type, there _has_ to be an allocation for the return value when there is a redaction -- whether it's done via StringBuilder or another hand-rolled implementation.

Reply:
> Who are you asking? 

Nobody in particular - I was just posing the question for discussion.

> Certainly, if the redacted text is known to consist of a single char, then we can further optimize (although, there's an edge case when a disallowed query param has a zero-length value -- nothing after the equals sign -- and the redacted string could still be longer than the input). 

I don't believe we insert the redacted placeholder in this case, so this shouldn't be a problem.

> However, without changing the return type, there _has_ to be an allocation for the return value when there is a redaction -- whether it's done via StringBuilder or another hand-rolled implementation.

There would still be allocation, but the logic would be significantly simpler.

Reply:
For what it's worth, I like this suggestion.

Reply:
OK, but not in this PR, right? Someone else should shepherd that idea to conclusion.